Robot goal addition

Robin27182 opened the pull request:
Added a RobotGoal class that holds a list of all of the presets we can have. It uses a builder pattern.
I was very tired of the 6 long named arguments in the constructors in choreoSubsystem, so I now have them taking in a RobotGoal instead.
In the RobotContainer, I made a teleGoal, (something like that), that houses our next wanted location to autonomously go to in Teleop. 
I made this a pull request because this is a drastic change, and I've done this without consultation. I think it's worked fairly well, I'm just glad I don't have 30 strings and horrible constructors.

Robin27182 committed:
Organized Subsystems into folders
Added RobotGoal class
THIS IS A BIG CHANGE. This needs to be reviewed
ChoreoSubsystem was super messy, this was my attempt at solving the problem.
Switching devices once again. This commit is unfinished. I will be actively working so I commit no more errors.

Robin27182 committed:
This should be the cocmpleted implementation of the addition of RobotGoal.
Made robotGoal mexh with choreo, being sent from the RobotContainer.
Used a builder pattern to assign states to go to via controller. SOME "Please don't explode" edge case code. Just please dont feed in empty robotGoals, make sure its filled all it needs to be. We don't need default things doing unexpected things.

clockworkthought commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -79,29 +90,101 @@ private void configureBindings() {
+        payloadController.y().whileTrue(choreo.liftToPos(teleGoal.copy().setLift(LiftPresets.MIDDLE_REEF)));
+        payloadController.b().whileTrue(choreo.liftToPos(teleGoal.copy().setLift(LiftPresets.TOP_REEF)));
+        payloadController.povDown().whileTrue(choreo.liftToPos(teleGoal.copy().setLift(LiftPresets.ALGAE_HIGH)));
+        payloadController.povUp().whileTrue(choreo.liftToPos(teleGoal.copy().setLift(LiftPresets.ALGAE_LOW)));
```
While this won't break anything as-implemented, you usually want to avoid calling the .copy() method unless you have a really good reason to make a new instance of a class with the same values as an existing instance.  You can either just pass the teleGoal in directly, and/or switch liftToPos() so that it accepts an argument of LiftPreset directly again (either as a sole option, or having liftToPos(RobotGoals) internally call liftToPos(LiftPreset)).  I'd favor doing the latter, but both are acceptable.

clockworkthought commented on src/main/java/frc/robot/subsystems/Autonomous/ChoreoSubsystem.java:
```diff
@@ -106,118 +120,91 @@ public Command getCoral() {
+    public Command alignToApril(LimelightSubsystem limelight, AprilAlign.AprilPositions aprilOffset) {
         return new SequentialCommandGroup(
-                new AprilAlign(driveSubsystem, limelightSubsystem, .42, aprilOffset),
+                new AprilAlign(driveSubsystem, limelight, .42, aprilOffset),
```
Probably want to use the aprilOffsets in your RobotGoal structure to represent this value.  0.42 is a pretty nice magic number, but the rear-facing camera ends up with an offset closer to 0.60.

clockworkthought commented on src/main/java/frc/robot/subsystems/Autonomous/RobotGoal.java:
```diff
@@ -0,0 +1,103 @@
+    private LiftPresets liftGoal;
+    private AlgaePresets algaeGoal;
+    private CoralPresets coralGoal;
+    private AprilAlign.AprilPositions aprilOffset;
```
I will strongly recommend giving these all safe default values.  We /shouldn't/ accidentally call an uninitialized variable, but in robotics you want to be absolutely sure.

clockworkthought reviewed (commented):
This approach looks good.  The RobotGoal class has a few ways we can polish it, and its implementation, but mostly nitpicks.  We'll work tonight on testing this implementation.

Robin27182 committed:
Fixed various problems as mentioned in the pull request:
Limelight now holds it's own flushOffset variable. It's a limelight to limelight basis, so that made the most sense to me
Although RobotGoal had a null check and default there, it was horribly unclear. Kept the "if null" implementation, and added default values to each variable.
Reworked the "reset" method in RobotGoal. It just returned a new instance, and didn't reset the current one. Was not very intuitive.
Removed usages of RobotGoal.copy because it is not yet needed.
Added possibilities for moveLiftTo to be given a goal or a preset directly.

Robin27182 commented:
Think I fixed the problems. More specific details in the commit message.
